Add Composite Check Functionality to SBOL Utilities

vinayakjeet opened the pull request:
#69 

## Overview
This PR introduces a new helper function, `is_composite`, to the SBOL Utilities package, addressing the need for an efficient way to determine if an SBOL Component is a composite. This functionality was proposed to tackle issue #69, focusing on components that are of DNA type and linked to an assembly plan via the `generated_by` attribute.

## Changes
- **`is_composite` Function:** Determines whether an SBOL Component is a composite by checking for DNA type and assembly plan linkage.
- **Supporting Functions:** Added `has_dna_type` and `has_assembly_plan` as helper functions to modularize the checks required for the `is_composite` function.

## Justification
The need for a composite check function was identified as a part of enhancing the SBOL utilities package to support more nuanced checks for components within SBOL documents. The ability to identify composite components efficiently is crucial for developers working with synthetic biology data, facilitating more sophisticated data analysis and manipulation tasks.

## Implementation Details
- The `has_dna_type` function utilizes the `tyto` library to map component types to known DNA terms, ensuring that the component is of DNA type.
- The `has_assembly_plan` function searches for the `generated_by` attribute in the component and verifies that any linked activities are marked with both `http://sbols.org/v3#assemblyPlan` and `sbol:design` types, adhering to SEP 054 specifications.
- The `is_composite` function integrates these checks, offering a straightforward way to determine if a component meets the criteria for being considered a composite.

## Future Work
While this PR addresses the immediate need for composite component identification, future enhancements could include:
- Expanding the criteria for what constitutes a composite component based on evolving standards.
- Optimizing performance for large SBOL documents through more efficient data structures or caching mechanisms.

## Acknowledgements
Special thanks to contributors and reviewers who have offered insights and feedback throughout the development of this feature, especially those involved in the discussion of issue #69.

closes #69

vinayakjeet committed:
update4

Signed-off-by: Vinayakjeet Singh Karki <139736674+vinayakjeet@users.noreply.github.com>

jakebeal commented:
Tests are failing, but currently only the two caused by #231, so the testing is sufficient to proceed to a full review.

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -363,4 +363,21 @@ def is_circular(obj: Union[sbol3.Component, sbol3.LocalSubComponent, sbol3.Exter
+        return any(tyto.SO.DNA == tyto.SO.get_uri_by_term(t) for t in o.types)
+
+    def has_assembly_plan(o):
+        if not hasattr(o, "generated_by"):
```
I don't think this should be necessary, because every `Component` has a generated_by attribute, even if it is empty.

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -363,4 +363,21 @@ def is_circular(obj: Union[sbol3.Component, sbol3.LocalSubComponent, sbol3.Exter
+            return False
+        for activity_ref in o.generated_by:
+            activity = activity_ref.lookup()
+            if activity and 'http://sbols.org/v3#assemblyPlan' in activity.types and 'sbol:design' in activity.types:
```
Please change `'sbol:design'` to the `SBOL_DESIGN` constant.

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -363,4 +363,21 @@ def is_circular(obj: Union[sbol3.Component, sbol3.LocalSubComponent, sbol3.Exter
+    return any(n==sbol3.SO_CIRCULAR for n in obj.types)
+ 
+def is_composite(obj):
+    def has_dna_type(o):
```
Function needs docstring

jakebeal reviewed (changes requested):
A few minor adjustments are needed.
More importantly, a test needs to be added that will validate that this function behaves as expected.

vinayakjeet committed:
Merge branch 'SynBioDex:develop' into update4

vinayakjeet committed:
Update helper_functions.py

Hi @jakebeal,  

I have implemented all requested changes:  

- **Removed** the unnecessary `hasattr(o, "generated_by")` check.  
- **Replaced** `'sbol:design'` with `sbol3.SBOL_DESIGN`.  
- **Added** a docstring to `is_composite()` for better documentation.  

Please review and let me know if any further modifications are needed. Thanks! 🚀

vinayakjeet committed:
Merge pull request #1 from vinayakjeet/vinayakjeet-patch-1

Update helper_functions.py

vinayakjeet committed:
Update helper_functions.py

Hi @jakebeal,  

I have implemented all requested changes:  

- **Removed** the unnecessary `hasattr(o, "generated_by")` check.  
- **Replaced** `'sbol:design'` with `sbol3.SBOL_DESIGN`.  
- **Added** a docstring to `is_composite()` for better documentation.  

Please review and let me know if any further modifications are needed. Thanks! 🚀

vinayakjeet commented:
Hi @jakebeal,  

Implemented all requested changes:  

- **Removed** the unnecessary `hasattr(o, "generated_by")` check.  
- **Replaced** `'sbol:design'` with `sbol3.SBOL_DESIGN`.  
- **Added** a docstring to `is_composite()`
 

Please review and let me know if any further modifications are needed. Thanks! 🚀

vinayakjeet committed:
Update test_helpers.py

#69 test cases

vinayakjeet commented:
Added test cases for composite component checks. Please review and run.

vinayakjeet commented:
@Gonza10V  @jakebeal  Can you please run the changes in the PR. Thanks🚀
